examples/performance-debugging/performance-debugging.cpp

```diff
@@ -0,0 +1,501 @@
+    // value `-1.0`.
+    void on_iteration_complete(const gko::LinOp *, const gko::size_type &,
+                               const gko::LinOp *residual,
+                               const gko::LinOp *solution,
+                               const gko::LinOp *residual_norm) const override
+    {
+        if (residual_norm) {
+            rec_res_norms.push_back(
+                utils::get_norm(gko::as<vec<ValueType>>(residual_norm)));
+        } else {
+            rec_res_norms.push_back(
+                utils::compute_norm(gko::as<vec<ValueType>>(residual)));
+        }
+        if (solution) {
+            true_res_norms.push_back(utils::compute_residual_norm(
+                matrix, b, gko::as<vec<ValueType>>(solution)));
+        } else {
+            true_res_norms.push_back(-1.0);
+        }
+    }
+
+    ResidualLogger(std::shared_ptr<const gko::Executor> exec,
+                   const gko::LinOp *matrix, const vec<ValueType> *b)
+        : gko::log::Logger(exec, gko::log::Logger::iteration_complete_mask),
+          matrix{matrix},
+          b{b}
+    {}
+
+    void write_data(std::ostream &ostream)
+    {
+        ostream << "Recurrent Residual Norms: " << std::endl;
+        ostream << "[" << std::endl;
+        for (const auto &entry : rec_res_norms) {
+            ostream << "\t" << entry << std::endl;
+        }
+        ostream << "];" << std::endl;
+
+        ostream << "True Residual Norms: " << std::endl;
+        ostream << "[" << std::endl;
+        for (const auto &entry : true_res_norms) {
+            ostream << "\t" << entry << std::endl;
+        }
+        ostream << "];" << std::endl;
+    }
+
+private:
+    const gko::LinOp *matrix;
+    const vec<ValueType> *b;
+    mutable std::vector<ValueType> rec_res_norms;
+    mutable std::vector<ValueType> true_res_norms;
+};
+
+
+}  // namespace loggers
+
+
+namespace {
+
+
+// Print usage help
+void print_usage(const char *filename)
+{
+    std::cerr << "Usage: " << filename << " [executor] [matrix file]"
+              << std::endl;
+    std::cerr << "matrix file should be a file in matrix market format. "
+                 "The file data/A.mtx is provided as an example."
+              << std::endl;
+    std::exit(-1);
+}
+
+
+void print_vector(const gko::matrix::Dense<> *vec)
+{
+    std::cout << "[" << std::endl;
+    for (int i = 0; i < min(10, vec->get_size()[0]); ++i) {
+        std::cout << "\t" << vec->at(i) << std::endl;
+    }
+    std::cout << "];" << std::endl;
+}
+
+
+}  // namespace
+
+
+int main(int argc, char *argv[])
+{
+    // Parametrize the benchmark here
+    // Pick a value type
+    using vtype = double;
+    // Pick a matrix format
+    using mtx = gko::matrix::Csr<vtype>;
+    // Pick a solver
+    using solver = gko::solver::Cg<vtype>;
+    // Pick a preconditioner type
+    using preconditioner = gko::matrix::IdentityFactory<vtype>;
+    // Pick a residual norm reduction value
+    auto reduction_factor = 1e-8;
```

Comment:
Do you think it would be difficult to add gflags support here so that the user can do this from the command line ?

Reply:
I pondered that at first, here are some of my thoughts.

This would complexify ginkgo by quite a bit for not much: gflags currently is given to the benchmarks, here we would need to add that dependency to examples only for that one example.

Secondly, since the purpose of this is to debug and also to learn about Ginkgo at the same time, I think adding the complexity of the flags system and already instantiating for the user all the preconditioners etc which you can select (as in done in the benchmarks) can be detrimental. You also lose touch with the fact that "adapting Ginkgo to your needs is easy" (should be). With the current format, you can tinker manually with the important parameters and start doing a little bit of coding in a hopefully simple manner, whereas passing flags only makes you learn about the example's interface but not the code itself.

Finally, the solver benchmark actually does all of that already pretty much exactly, so in fact if the purpose really is to only try a lot of combinations without any coding/user involvement then simply going there and using the proper tool should be the better solution.

Reply:
Hm, okay. That makes sense.